### PR TITLE
Pod tasks can be used for map tasks

### PR DESCRIPTION
**Pull request by katrogan**
# TL;DR
Pod tasks can now be mapped over too. As a long term goal we should introduce Array Nodes as proposed [here](https://github.com/flyteorg/flyteidl/pull/178) but this PR is an interim solution given how much refactoring introducing the new node type entails.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1051

## Follow-up issue
_NA_

### CODE REVIEW

File: go/tasks/plugins/array/k8s/task.go [hunk 1]
```diff
@@ -60,33 +81,36 @@ func (t Task) Launch(ctx context.Context, tCtx core.TaskExecutionContext, kubeCl
...
+	args := append(podTemplate.Spec.Containers[containerIndex].Command, podTemplate.Spec.Containers[containerIndex].Args...)
+	podTemplate.Spec.Containers[containerIndex].Command = []string{}
 
 	indexStr := strconv.Itoa(t.ChildIdx)
 	podName := formatSubTaskName(ctx, tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName(), indexStr)
 
 	pod := podTemplate.DeepCopy()
 	pod.Name = podName
-	pod.Spec.Containers[0].Env = append(pod.Spec.Containers[0].Env, corev1.EnvVar{
+	pod.Spec.Containers[containerIndex].Env = append(pod.Spec.Containers[0].Env, corev1.EnvVar{
```
Comment (EngHabu) [inline go/tasks/plugins/array/k8s/task.go:100]:
```suggestion
	pod.Spec.Containers[containerIndex].Env = append(pod.Spec.Containers[containerIndex].Env, corev1.EnvVar{
```

File: go/tasks/plugins/array/k8s/transformer.go [hunk 2]
```diff
@@ -49,6 +50,48 @@ func GetNamespaceForExecution(tCtx core.TaskExecutionContext, namespaceTemplate
 	return namespace
 }
 
+// Initializes a pod from an array job task template with a K8sPod set as the task target.
+// TODO: This should be removed (it duplicates the pod plugin logic) once we improve array job handling
```
Comment (EngHabu) [inline go/tasks/plugins/array/k8s/transformer.go:54]: nit: add a date here... let's see when will we actually get to this... any bets?

File: go/tasks/plugins/array/k8s/transformer.go [hunk 3]
```diff
@@ -99,6 +137,22 @@ func FlyteArrayJobToK8sPodTemplate(ctx context.Context, tCtx core.TaskExecutionC
...
-	}, arrayJob, nil
+	}
+	if taskTemplate.GetContainer() != nil {
+		podSpec, err := flytek8s.ToK8sPodSpec(ctx, arrTCtx)
+		if err != nil {
+			return v1.Pod{}, nil, err
+		}
+		pod.Spec = *podSpec
+	} else if taskTemplate.GetK8SPod() != nil {
+		k8sPod, err := buildPodMapTask(taskTemplate, tCtx.TaskExecutionMetadata())
```
Comment (EngHabu) [inline go/tasks/plugins/array/k8s/transformer.go:148]: How does this resolve command line args in the primary container?
`pyflyte-execute --inputsFile {{.inputs}}`
where does that get resolved? it's not clear to me...

Comment (katrogan) [reply go/tasks/plugins/array/k8s/transformer.go:148]: done here: https://github.com/flyteorg/flyteplugins/pull/182/files#diff-01ee392f44bbd91e1e5f4afd999327e02f8f1df54fd5b5b576478e903cbcb3c2R113

File: go/tasks/plugins/array/k8s/task.go (hunk 1 above)
Comment (katrogan) [reply go/tasks/plugins/array/k8s/task.go:100]: yikes, thank you for catching!

File: go/tasks/plugins/array/core/state.go [hunk 4]
```diff
@@ -276,7 +276,7 @@ func SummaryToPhase(ctx context.Context, minSuccesses int64, summary arraystatus
 	// No chance to reach the required success numbers.
 	if totalRunning+totalSuccesses+totalWaitingForResources < minSuccesses {
 		logger.Infof(ctx, "Array failed early because total failures > minSuccesses[%v]. Snapshot totalRunning[%v] + totalSuccesses[%v] + totalWaitingForResource[%v]",
-			totalRunning, totalSuccesses, totalWaitingForResources, minSuccesses)
+			minSuccesses, totalRunning, totalSuccesses, totalWaitingForResources)
```
[the remark this answers is not in the document]
Comment (katrogan) [reply go/tasks/plugins/array/core/state.go:279]: made for some head scratching debugging, heh

File: go/tasks/plugins/array/k8s/transformer.go (hunk 2 above)
Comment (katrogan) [reply go/tasks/plugins/array/k8s/transformer.go]: done, let me know if you have other suggestions :)

File: go/tasks/plugins/array/k8s/task.go [hunk 5]
```diff
@@ -60,33 +81,36 @@ func (t Task) Launch(ctx context.Context, tCtx core.TaskExecutionContext, kubeCl
...
+	pod.Spec.Containers[containerIndex].Env = append(pod.Spec.Containers[containerIndex].Env, arrayJobEnvVars...)
 	taskTemplate, err := tCtx.TaskReader().Read(ctx)
 	if err != nil {
 		return LaunchError, errors2.Wrapf(ErrGetTaskTypeVersion, err, "Unable to read task template")
 	} else if taskTemplate == nil {
 		return LaunchError, errors2.Wrapf(ErrGetTaskTypeVersion, err, "Missing task template")
 	}
 	inputReader := array.GetInputReader(tCtx, taskTemplate)
-	pod.Spec.Containers[0].Args, err = template.Render(ctx, args,
+	pod.Spec.Containers[containerIndex].Args, err = template.Render(ctx, args,
```
Comment (EngHabu) [inline go/tasks/plugins/array/k8s/task.go:113]: I guess this should apply to cmd as well?

File: go/tasks/plugins/array/k8s/transformer.go (hunk 3 above)
Comment (EngHabu) [reply go/tasks/plugins/array/k8s/transformer.go:148]: Aha, I guess we should move https://github.com/flyteorg/flyteplugins/pull/182/files#diff-f6e976bede1900c30812a8f3dc9c733a2a60381b4ad8e38fe33f0525bf0bb118R113-R116 into the `if container` condition above?

File: go/tasks/plugins/array/k8s/task.go (hunk 5 above)
Comment (katrogan) [reply go/tasks/plugins/array/k8s/task.go:113]: we already combine the command and args: https://github.com/flyteorg/flyteplugins/pull/182/files#diff-01ee392f44bbd91e1e5f4afd999327e02f8f1df54fd5b5b576478e903cbcb3c2R92

(for the record, primary container always has an sdk supplied command + args)

File: go/tasks/plugins/array/k8s/transformer.go (hunk 3 above)
Comment (katrogan) [reply go/tasks/plugins/array/k8s/transformer.go:148]: @EngHabu not sure - do we allow for input substitution with the custom?

Comment (katrogan) [reply go/tasks/plugins/array/k8s/transformer.go:150]: updated @EngHabu

**Comment by katrogan** [comment]
PTAL @EngHabu

File: go/tasks/plugins/array/k8s/task.go (hunk 5 above)
Comment (EngHabu) [reply go/tasks/plugins/array/k8s/task.go:113]: I don't think, generally speaking, we should make assumptions based on how we implemented Flytekit... the underlying spec allows cmd & args and I don't think this plugin should change that behavior... What do yo think?

File: go/tasks/plugins/array/k8s/transformer.go [hunk 6]
```diff
@@ -99,6 +139,39 @@ func FlyteArrayJobToK8sPodTemplate(ctx context.Context, tCtx core.TaskExecutionC
...
+		pod.Annotations = utils.UnionMaps(pod.Annotations, k8sPod.Annotations)
+		pod.Spec = k8sPod.Spec
+
+		// Here we templatize the k8sPod primary container args. The call to ToK8sPodSpec for the task container target
+		// case already handles this but we must explicitly do so for K8sPod task targets.
+		containerIndex, err := getTaskContainerIndex(&pod)
+		if err != nil {
+			return v1.Pod{}, nil, err
+		}
+		pod.Spec.Containers[containerIndex].Args, err = template.Render(ctx, pod.Spec.Containers[containerIndex].Args,
```
Comment (EngHabu) [inline go/tasks/plugins/array/k8s/transformer.go:164]: I think we should do this on `Command` as well... like we do in `ToK8sPod`

File: go/tasks/plugins/array/k8s/task.go (hunk 5 above)
Comment (EngHabu) [reply go/tasks/plugins/array/k8s/task.go:113]: Also, I thought we are getting rid of the args/cmd manipulation altogether from this function, right?

File: go/tasks/plugins/array/k8s/transformer.go (hunk 6 above)
Comment (katrogan) [reply go/tasks/plugins/array/k8s/transformer.go:164]: done

File: go/tasks/plugins/array/k8s/task.go (hunk 5 above)
Comment (katrogan) [reply go/tasks/plugins/array/k8s/task.go]: oh derp, yes. thanks & done

**Review by EngHabu — APPROVED** [review]
